Split into declaration and assignment issue for var type

vikasprabhakar opened the pull request:
Reference jira bug : https://issues.apache.org/jira/browse/NETBEANS-465

Issue Details:
Warning hint "Split into declaration and assignment" is not valid for var type.
var v = 10;
Only declaration the var type variable causes the errors:
var v; -> cannot infer type for local variable v
   (cannot use 'var' on variable without initializer)

The hints "Split into declaration and assignment" should not show for var type declaration and assignments

rtaneja1 commented:
1 Are there any other .Kind for which control comes here?

2 Why startsWith check? Can we check for equals here.

3 If we are using any constant strings, we should add a line comment //NOI18N to skip flagging by nls audit tool.

vikasprabhakar commented:
Added rtaneja1 review comment as:
1 Are there any other .Kind for which control comes here?
--> Remove the kind check
2 Why startsWith check? Can we check for equals here.
--> Can't use code.equals("var "), because code has full statement as String.
3 If we are using any constant strings, we should add a line comment //NOI18N to skip flagging by nls audit tool.
--> Implemented

jlahoda commented:
My personal opinions:
-no need for a comment referring to jira, adding such a comment for every change would lead to many comments. The VCS log should be enough, and tests should help as well.
-there should be a (unit) test. Tests by default run with nb-javac, and nb-javac supports vars, so this should be doable, even when running the tests with JDK 8. The test may refer to the Jira #, so that if it fails, one can find out the original report.
-please avoid looking at the text as much as possible - such things are quite fragile (e.g., what if someone writes "var/**/ v = 1;"? Not sure if the fact that the variable is a "var" can be detected from the AST, maybe not. But it should be possible to use lexer tokens to check for "var", something like CompilationInfo.getTreeUtilities().tokensFor(VariableTree) and then check the appropriate token. Please note that things like this are allowed:
        final var v1 = 0;
        @Deprecated var v2 = 1;
So I'd suggest to move the token sequence behind the modifiers, ignore whitespaces/comments and check the following token.
-I'd suggest to do an API method for this check in TreeUtilities. I suspect this check will be needed on more places, so better not to duplicate it.

sdedic commented:
I haven't checked, but VariableTree.type's position may be -1 in this case. In that case check also in syntax with multiple declarators like `int a,b;`

vikasprabhakar commented:
Thanks sdedic for your valuable feedback.
1. I added enum check instead of equals method.
2. I added the condition to check the var keyword in only declaration side and not in assignment side.
Please find the test case testSplitDeclarationForVar2() for your mentioned test scenario.
Let me know if you need any inputs.

vikasprabhakar commented:
Method added in utilities class

jlahoda commented on java.source.base/src/org/netbeans/api/java/source/TreeUtilities.java:
```diff
@@ -1850,7 +1850,23 @@ public boolean isCompileTimeConstantExpression(TreePath expression) {
+     * @param leaf the leaf must be {@link Tree}
+     * @return the true if tree contains var keyword else return false.
+     */
+    public boolean isVarKeywordAvailable(@NonNull Tree leaf) {
```
Please rename the method to say the intent what it is doing, not how - i.e. "isVarVariable" or something like that. The point of having this method is that we can change the implementation if a better way is found. Add @since and increment spec.version.base (currently is 2.30.0, should be 2.31.0) in java.source.base/nbproject/project.properties. The parameter should either be VariableTree or a TreePath.

jlahoda commented on java.hints/src/org/netbeans/modules/java/hints/suggestions/Tiny.java:
```diff
@@ -222,8 +222,12 @@ public static ErrorDescription splitDeclaration(HintContext ctx) {
+        }      
 
-        String displayName = NbBundle.getMessage(Tiny.class, "ERR_splitDeclaration");
+	String displayName = NbBundle.getMessage(Tiny.class, "ERR_splitDeclaration");
```
Please undo this noop change (the indent was better before). Thanks.

junichi11 commented on java.hints/src/org/netbeans/modules/java/hints/suggestions/Tiny.java:
```diff
@@ -222,8 +222,12 @@ public static ErrorDescription splitDeclaration(HintContext ctx) {
 
         if (parentKind != Tree.Kind.BLOCK && parentKind != Tree.Kind.CASE) return null;
+        
+	if(ctx.getInfo().getTreeUtilities().isVarKeywordAvailable(ctx.getPath().getLeaf())){
```
Please don't use TAB.

sdedic commented:
Specification version changed, apichanges.xml should mention the public API change.

geertjanw commented:
Discussed with Jan Lahoda, myself, Vikas -- we're OK with this and merging.
